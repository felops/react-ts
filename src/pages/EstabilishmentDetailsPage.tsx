import { useNavigate, useParams } from 'react-router';
import { useQuery } from "@tanstack/react-query";
import { getEstablishmentRatingsById } from "../api/ratingsAPI";

export const EstablishmentDetailsPage = () => {
  const { id } = useParams();
  const navigate = useNavigate();
  const { isLoading, data, error } = useQuery({
    queryKey: ['establishments', id],
    queryFn: () => getEstablishmentRatingsById(id),
    enabled: Boolean(id)
  });

  return (
    <>
      <button onClick={() => navigate(-1)}>go back</button>
      <h2>Establishment Details</h2>
      {
        isLoading
          ? <p>Loading...</p>
          : (
            <div>
              <p>Date of inspection: {data?.RatingDate ? new Date(data?.RatingDate).toLocaleDateString() : '-'}</p>
              <p>Rating: {data?.RatingValue}</p>
              <p>Adress:</p>
              <p>{data?.AddressLine1}</p>
              <p>{data?.AddressLine2}</p>
              <p>{data?.AddressLine3}</p>
              <p>{data?.AddressLine4}</p>
            </div>
          )
      }
    </>
  );
};

export default EstablishmentDetailsPage;
