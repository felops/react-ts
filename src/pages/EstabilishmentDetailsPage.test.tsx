import { fireEvent, render } from '@testing-library/react';
import { useQuery } from "@tanstack/react-query";
import { EstablishmentDetailsPage } from './EstabilishmentDetailsPage';
import { BrowserRouter } from 'react-router';
import * as ReactRouter from 'react-router';

const mockData = {
  RatingDate: new Date('2000-01-01T12:00:00Z'),
  RatingValue: "5",
  AddressLine1: "Coronation Street",
  AddressLine2: "Hide Park",
  AddressLine3: "New York",
  AddressLine4: "MC",
}

jest.mock('@tanstack/react-query', () => ({
  useQuery: jest.fn(),
}));

const mockUseNavigate = jest.fn();
jest.mock('react-router', () => ({
   ...jest.requireActual('react-router'),
  useNavigate: () => mockUseNavigate,
}));

describe("EstabilishmentDetailsPage", () => {
  it('renders correctly when is loading', () => {
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: true,
      data: mockData,
      error: null,
    });
    const { asFragment } = render(
      <BrowserRouter>
        <EstablishmentDetailsPage />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when is NOT loading', () => {
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      data: mockData,
      error: null,
    });
    const { asFragment } = render(
      <BrowserRouter>
        <EstablishmentDetailsPage />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should call navigate function when button is clicked', () => {
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      data: mockData,
      error: null,
    });
    render(
      <BrowserRouter>
        <EstablishmentDetailsPage />
      </BrowserRouter>
    );
    fireEvent.click(document.querySelector('button')!);
    expect(mockUseNavigate).toHaveBeenCalledWith(-1);
  });
});
