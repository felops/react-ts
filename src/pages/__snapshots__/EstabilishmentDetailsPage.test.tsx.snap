// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EstabilishmentDetailsPage renders correctly when is NOT loading 1`] = `
<DocumentFragment>
  <button>
    go back
  </button>
  <h2>
    Establishment Details
  </h2>
  <div>
    <p>
      Date of inspection: 1/1/2000
    </p>
    <p>
      Rating: 5
    </p>
    <p>
      Adress:
    </p>
    <p>
      Coronation Street
    </p>
    <p>
      Hide Park
    </p>
    <p>
      New York
    </p>
    <p>
      MC
    </p>
  </div>
</DocumentFragment>
`;

exports[`EstabilishmentDetailsPage renders correctly when is loading 1`] = `
<DocumentFragment>
  <button>
    go back
  </button>
  <h2>
    Establishment Details
  </h2>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;
